coins/monero/src/ringct/mod.rs

```diff
@@ -10,24 +10,54 @@ pub use hash_to_point::{raw_hash_to_point, hash_to_point};
 
 /// CLSAG struct, along with signing and verifying functionality.
 pub mod clsag;
+/// MLSAG struct, along with verifying functionality.
+pub mod mlsag;
+/// RangeSig struct, along with verifying functionality.
+pub mod borromean;
 /// Bulletproofs(+) structs, along with proving and verifying functionality.
 pub mod bulletproofs;
 
 use crate::{
   Protocol,
   serialize::*,
-  ringct::{clsag::Clsag, bulletproofs::Bulletproofs},
+  ringct::{clsag::Clsag, mlsag::MgSig, bulletproofs::Bulletproofs, borromean::RangeSig},
 };
 
 /// Generate a key image for a given key. Defined as `x * hash_to_point(xG)`.
 pub fn generate_key_image(secret: &Zeroizing<Scalar>) -> EdwardsPoint {
   hash_to_point(&ED25519_BASEPOINT_TABLE * secret.deref()) * secret.deref()
 }
 
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum EcdhInfo {
+  Standard { mask: Scalar, amount: Scalar },
+  Bulletproof { amount: [u8; 8] },
+}
+
+impl EcdhInfo {
+  pub fn read<R: Read>(rct_type: u8, r: &mut R) -> io::Result<EcdhInfo> {
+    Ok(match rct_type {
+      0 ..= 3 => EcdhInfo::Standard { mask: read_scalar(r)?, amount: read_scalar(r)? },
```

Comment:
If we keep using rct_type like this, they should be turned into proper constants.

I'd rather have this take in a boolean of which to use though, to prevent distribution of such specificities.